### PR TITLE
gui: Show internal config and state paths (fixes #8323)

### LINKED ISSUE #8323: Show internally used paths in the GUI for debugging
### Include required information

With the trouble some people have recently had regarding some third-party packages (like Synology), it seems that the path to config and database is sometimes hard to find out.  The package integrations configure some exotic path as home directory for example, which needs to be known for configuration backups or for running the CLI.

The info is already available through the `--paths` command line switch, but only the default values and only when supplying the correct home directory in the first place.

It should be possible to query that info from a running instance through the API and within the GUI.

### PR DESCRIPTION
**Pull request by acolomb**
### Purpose

Make the information shown by the `--paths` command line option accessible through the GUI, under a new tab in the Logs modal.  Should help e.g. finding the config file on exotic platforms that set some unexpected home directory through the package integrations.

This is exposed through a new API endpoint, together with some cleanup in the `locations` package.  Some more detailed explanations to be found in the commit messages.

### Testing

No new tests added.

### Screenshots

image

### Documentation

* [x] REST API extension see https://github.com/syncthing/docs/pull/755
* [ ] Possibly a hint somewhere in the troubleshooting section?

### CODE REVIEW

**Comment by acolomb** [comment]
Some ideas that didn't make it into this proposal: 

* Extending the `Starting` event to show all paths besides the `home` directory.  Theoretically sufficient to announce this once, but doesn't help if the GUI is opened after the backend.
* Adding a `paths` attribute to the `/rest/system/status` endpoint instead.
* Adding descriptive titles for each entry as translations for the `LocationEnum` keys. Didn't seem worth the trouble as this is mainly needed to diagnose technical problems.  So a simple `uncamel` seems good enough.
* Other places where to hide this info in the GUI.  Suggestions welcome!

File: gui/default/syncthing/core/logViewerModalView.html [hunk 1]
```diff
@@ -26,6 +27,18 @@
         </table>
       </div>
 
+      <div id="log-viewer-paths" class="tab-pane">
+        <label translate>Internally used paths:</label>
+        <table class="table table-striped">
+          <tbody>
+            <tr ng-repeat="(name, path) in logging.paths">
+              <td class="text-nowrap">{{ name | uncamel }}</td>
+              <td><code>{{ path }}</code></td>
```
Comment (tomasz1986) [inline gui/default/syncthing/core/logViewerModalView.html:36]:
```suggestion
              <td class="file-path"><code>{{ path }}</code></td>
```

The GUI is broken on small screens due to wrapping not being allowed.

Before:

image

After:

image

**Comment by calmh** [comment]
This is a great addition, long overdue. However, I think we should do it a bit more manually, and the `--paths` output is maybe not a great model. Looking at each of the items listed,

- csrf tokens: never useful to the user
- cert file, key file, https cert file, https key file: potentially useful, but always live in the config directory
- database: potentially useful, can be separate from the config directory
- default folder path: never useful, I think
- gui assets: potentially useful (though always in the config directory)
- log file: certainly useful

So I think we could boil this down to showing "Configuration directory (contains configuration and keys)", "Database location" (the database is a directory, but that's kind of an implementation detail), "GUI overrides directory" (or something like that) and "Log file", each with localization...

**Comment by acolomb** [comment]
Thanks for your comments, I'm fine with going that route in general.

> gui assets: potentially useful (though always in the config directory)

Not quite true, `STGUIASSETS` can be set to anything.

> So I think we could boil this down to showing "Configuration directory (contains configuration and keys)", [...]

I don't think hard-coding that assertion into the GUI code is a good idea.  That's the last place one would think of adjusting when something like the conf / data split ever happens again.  Better print full paths to each item, that's safer for the situations I'm thinking about where people are already in trouble because of something.  Combining a directory path with a file name might be just one bit too much to handle correctly.

My primary goal was to mirror the `--paths` output in machine readable form, but UI wise I guess we can streamline it and possibly improve `--paths` output as well to match the GUI.

**Comment by imsodin** [comment]
Looks good. I'd put it into it's own modal/actions menu item - under logs it's not very discoverable. Or maybe putting it in "about", that'd make more sense to me than logs (has information on the build already).

**Comment by acolomb** [comment]
I had thought about the "About" dialog, but found it to have an overwhelming amount of information already.  I don't prefer making an own modal for two reasons: 1) not cluttering the Actions menu further with very seldomly used, but prominent stuff and 2) it's such a small piece of info that it just doesn't warrant a separate dialog.  The idea behind putting it in the Logs modal was that most people will need it during troubleshooting, where the first thing they will be sent to check are the logs.

Other opinions on this?

**Comment by tomasz1986** [comment]
> Other opinions on this?

Maybe rename "Logs" to "Troubleshooting" and then put it there (as this PR already does)?

The current name of the modal, i.e. "Logs", isn't exactly correct right now already, as it contains both logs and debugging, so I think it'd fit well. Just my opinion, of course 🙂.

**Comment by bt90** [comment]
I might be murdering someones baby for asking but can't we shrink the current About page? The list of contributors could be maintained on the Github readme for example.

**Comment by imsodin** [comment]
> The current name of the modal, i.e. "Logs", isn't exactly correct right now already, as it contains both logs and debugging, so I think it'd fit well. Just my opinion, of course slightly_smiling_face.

Well "Debugging" there is strictly about logging debug facilities, so that fits well in my opinion.

I don't think the action menu is very cluttered. If we want to keep the number of entries down, then a proposed "Troubleshooting" that replaces/incorporates the "Support Bundle" would make most sense to me. "Logs" is likely to be often used already, removing it might cause confusion. I think "Support Bundle" is niche enough that replacing it is ok.

**Comment by acolomb** [comment]
I like the idea of replacing the "Support Bundle" link. However, it took me some time to even find information about it, because it is hidden unless the Debugging flag ist turned on in advanced configuration. That's too much of a hurdle for accessing the paths.

**Comment by calmh** [comment]
The support bundle thing is specific functionality that is only available when debugging is enabled, yeah, not really related.

Frankly I think "About" is a good place. (I also think we should keep the list of all contributors there. Scrolling is alright.)

**Comment by tomasz1986** [comment]
> Well "Debugging" there is strictly about logging debug facilities, so that fits well in my opinion.
> 
> I don't think the action menu is very cluttered. If we want to keep the number of entries down, then a proposed "Troubleshooting" that replaces/incorporates the "Support Bundle" would make most sense to me. "Logs" is likely to be often used already, removing it might cause confusion. I think "Support Bundle" is niche enough that replacing it is ok.

The reason for suggesting the merge was that I think everything will become confusing with a new and separate "Troubleshooting" entry. We will have both "Logs" and "Troubleshooting" in the Actions menu, but then inside "Logs" there is "Debugging" whose purpose is also to troubleshoot (just the logs, but still). On top of that, logging in itself is commonly considered a part of the troubleshooting process too.

**Comment by acolomb** [comment]
My gut feeling from the discussion was that the About modal would have the best chances for a consensus.  Along the way, I reduced the necessary scrolling there by splitting the authors and third-party software lists into separate tabs, with the paths overview being the third tab.  Here goes:
Screenshot 2022-05-07 at 00-54-24 devtest Syncthing
Screenshot 2022-05-07 at 00-54-51 devtest Syncthing
Screenshot 2022-05-07 at 00-55-10 devtest Syncthing

**Comment by tomasz1986** [comment]
How is the paths order determined? I mean, the log file location is probably way more significant to the user than the HTTPS certificate one :wink:.

The order in the original commit was alphabetical, which I personally liked, but it can obviously be manual too. Just curious what each path's importance has been based on.

**Comment by acolomb** [comment]
I tried to apply some grouping as calmh suggested: config dir and everything below, then database, then everything that can be placed totally independently from those two paths. The user home directory is usually parent to the other two, so that goes right on top. Might help to clarify the environment, such as on Synology where `$HOME` needs to be set up manually when starting the service.

**Comment by tomasz1986** [comment]
Makes sense 🙂. Personally, I'd just like to suggest maybe adding an explanation when the log file is `-`, because this means absolutely nothing to a new user (or really any user, who hasn't studied the docs thoroughly). Maybe add a help link to https://docs.syncthing.net/users/syncthing.html?highlight=logfile#cmdoption-logfile if the log file equals `-`? If it's a full path, then no additional explanation is necessary though.

**Comment by acolomb** [comment]
I'd say `-` is pretty clear for a field labeled "Log **File**".  In the case where it displays `-`, there is no file.

**Comment by acolomb** [comment]
The comment on https://github.com/syncthing/syncthing/commit/520ca4bcb04574ec01ff69f2e2385160b5b87486 was valid IMHO and there was a small issue there as well with the `</div>` being indented wrongfully on author list updates.  In order not to cause conflicts, I included those fixes in this PR as well.

Not quite sure if the last commit is correct, or if the bot names will always come back to the `AUTHORS` file anyway?

**Comment by tomasz1986** [comment]
> I'd say `-` is pretty clear for a field labeled "Log **File**". In the case where it displays `-`, there is no file.

Yeah, that's what you think people will do, but in reality a non-techy person will see `-` and say: "What's this? What do you mean by a minus sign?" and then they will go on asking "Where's the log file then?", and so on. Just speaking from experience having helped literally a hundred or more people with their computer problems. Especially if it's a Windows user with zero command line experience, there is just no way that they will connect the dots and view the minus sign as "no log file" :sweat_smile:.

**Comment by acolomb** [comment]
I respectfully disagree, and we should put this into perspective.  It's quite usual to just strike out a field in a paper form for example if it's intentionally left blank because it does not apply.  And I see this dialog mostly useful during support when someone on the forums tells the user to go looking at it or post a screenshot (which we usually request anyway).  No newbie is expected to discover this by themselves.  And, well, if it doesn't give you a path, there is no path to look at, period.  IMHO it's not a big enough problem to justify adding more complexity to this code, such as conditionally adding documentation links.  Also the semantics of showing a dash here could change sometime in the distant future, and updating this piece of code is the last thing people will remember, unfortunately.

**Comment by tomasz1986** [comment]
> It's quite usual to just strike out a field in a paper form for example if it's intentionally left blank because it does not apply. I don't see a connection to the UNIX habit of `-` meaning standard input / output as a CLI argument.

Personally, I'm quite certain that the non-techy user will just view `-` as a random character with no clear meaning here. The CLI argument is important, because many popular command line tools use `-` in a similar manner (e.g. Wget), but someone who's never used CLI in their life will have no idea about it. Obviously, this applies to a vast majority of Windows (and probably Mac) users.

This isn't something critical though, so if you don't feel like adding some additional code here for this purpose, don't let this stop proceeding with the PR. I just like to provide at least some clues here and there for the newbie users, as the GUI is already very overwhelming for them.

**Comment by acolomb** [comment]
> I just like to provide at least some clues here and there for the newbie users, as the GUI is already very overwhelming for them.

That's certainly appreciated :-)  I'm just trying to balance code complexity vs. pedantic UI optimization for inexperienced users.  And as with most optimizations, time spent on them should be roughly proportional to the frequency at which the feature will get used.  So while I think this is worthwhile and potentially very helpful (such as when I deal with Synology support questions), it's also a place where hopefully very few users will ever need to go looking.  Hence the scale balances towards minimal code complexity here.

**Comment by calmh** [comment]
In the spirit of the above bikeshedding, why are we bolding the database directory (and log file if there isn't one)? We should absolutely mention where the database lives, but I think I can count the number of times I've wanted to manually poke around in there on my actual physical digits, and I've been doing this for almost a decade. :) It doesn't seem worthy of the emphasis.

**Comment by acolomb** [comment]
The emphasis was meant to provide a bit of visual separation which entries are child paths below others, but obviously it doesn't work well.  Removed except for the config directory, assuming you were okay with that.

**Comment by tomasz1986** [comment]
I've just tested the code in the current iteration. Is there any special reason not to allow wrapping in the first column? I didn't mention this before, because the original text was quite short, but now it's got longer (and may be even longer when translated to some languages), and this is how it looks on small screens.

1. Wrapping disabled

image

2. Wrapping enabled

image

**Comment by acolomb** [comment]
How did you enable it? I don't quite remember, but there was something about making the first column shrink to only the needed width that brought the non-wrapping behavior with it. How does it look on a wider screen with wrapping enabled? The intention was to give the path as much width as possible before wrapping the first column.

**Comment by tomasz1986** [comment]
> How did you enable it? I don't quite remember, but there was something about making the first column shrink to only the needed width that brought the non-wrapping behavior with it. How does it look on a wider screen with wrapping enabled? The intention was to give the path as much width as possible before wrapping the first column.

I was just playing around with the browser's developer tools, but having checked the code now, there doesn't seem to be a straightforward way to do it.

There's

https://github.com/syncthing/syncthing/blob/a162e8d9f926f3bcfb5deb7f9abbb960d40b1f5b/gui/default/assets/css/overrides.css#L423-L428

in the code for narrower resolutions, and I've got a feeling that it could/should probably be applied to all tables universally instead of just `table-condensed` anyway, but doing so would require thorough testing to verify that nothing else in the GUI breaks.

A hacky solution is to just apply the `file-path` class to both columns. The browser will then wrap them automatically when needed. It doesn't really change anything on larger screens unless the paths are extremely long.

**Comment by acolomb** [comment]
I think I have a nice solution, just adding the `table-auto` class should suffice.  It's the same logic as for the folder / device detail panels, allowing wrapping in the row header cell only for narrow screens.  This should not change any other usage AFAICS.

**Comment by tomasz1986** [comment]
Just tested. All looks fine to me 🙂.

Not sure whether the "heart" icon really matches the content after the changes, but that's probably a topic for a different discussion. Maybe, if https://github.com/syncthing/syncthing/pull/8349 is accepted, we could just use the "syncthing" icon instead?

File: lib/locations/locations.go [hunk 2]
```diff
@@ -104,11 +103,18 @@ var locationTemplates = map[LocationEnum]string{
...
 
 var locations = make(map[LocationEnum]string)
 
+// Override allows to change only specific, usually user-supplied paths.
+func Override(location LocationEnum, path string) {
+	switch location {
+	case LogFile, GUIAssets:
+		locations[location] = path
+	}
+}
```
Comment (imsodin) [inline lib/locations/locations.go:116]: It seems error prone to make it silently do nothing for everything except those two locations. Why not do it for any `location` value?

File: lib/locations/locations.go [hunk 3]
```diff
@@ -128,6 +134,43 @@ func expandLocations() error {
...
+		if strings.Contains(path, "${timestamp}") {
+			continue
+		}
+		res[string(key)] = path
+	}
+	return res
+}
+
+// PrettyPrintPaths returns a nicely formatted, human-readable listing
+func PrettyPrintPaths() string {
```
Comment (imsodin) [inline lib/locations/locations.go:160]: `Print` suggest output, not returning a string to me - I'd name it just `PrettyPaths()`.

File: lib/locations/locations.go [hunk 4]
```diff
@@ -128,6 +134,43 @@ func expandLocations() error {
...
+		fmt.Sprintf("Configuration file:\n\t%s\n\n", Get(ConfigFile)),
+		fmt.Sprintf("Device private key & certificate files:\n\t%s\n\t%s\n\n", Get(KeyFile), Get(CertFile)),
+		fmt.Sprintf("GUI / API HTTPS private key & certificate files:\n\t%s\n\t%s\n\n", Get(HTTPSKeyFile), Get(HTTPSCertFile)),
+		fmt.Sprintf("Database location:\n\t%s\n\n", Get(Database)),
+		fmt.Sprintf("Log file:\n\t%s\n\n", Get(LogFile)),
+		fmt.Sprintf("GUI override directory:\n\t%s\n\n", Get(GUIAssets)),
+		fmt.Sprintf("CSRF tokens file:\n\t%s\n\n", Get(CsrfTokens)),
+		fmt.Sprintf("Default sync folder directory:\n\t%s\n\n", Get(DefFolder)),
+	}
+	return strings.Join(lines, "")
```
Comment (imsodin) [inline lib/locations/locations.go:171]: I think using `strings.Builder` would be a bit nicer/simpler.

File: script/authors.go [hunk 5]
```diff
@@ -117,7 +117,7 @@ func main() {
 
 	authorsRe := regexp.MustCompile(`(?s)id="contributor-list">.*?</div>`)
 	bs := readAll(htmlFile)
-	bs = authorsRe.ReplaceAll(bs, []byte("id=\"contributor-list\">\n"+replacement+"\n    </div>"))
+	bs = authorsRe.ReplaceAll(bs, []byte("id=\"contributor-list\">\n"+replacement+"\n          </div>"))
```
Comment (imsodin) [inline script/authors.go:120]: Just checking: Is this change on purpose?

File: lib/locations/locations.go [hunk 6]
```diff
@@ -128,6 +134,43 @@ func expandLocations() error {
 	return nil
 }
 
+// ListExpandedPaths returns a machine-readable mapping of the currently
+// configured locations.  Unexpanded entries are filtered out.
```
Comment (imsodin) [inline lib/locations/locations.go:138]: What's the reason to filter those? Plus that shouldn't ever happen, right? (expansion happens at init time)

File: lib/locations/locations.go [hunk 7]
```diff
@@ -128,6 +134,43 @@ func expandLocations() error {
...
+func ListExpandedPaths() map[string]string {
+	res := make(map[string]string, len(locations))
+	for key, path := range baseDirs {
+		res["baseDir-"+string(key)] = path
```
Comment (imsodin) [inline lib/locations/locations.go:142]: I'd prefer two maps for base directories and paths/files over this prefix.

**Review by imsodin — COMMENTED** [review]
Adding an override for locations is a bit problematic, respectively it's just the first time where the pre-existing design with global variables shows problems:  
Right now everything is set up at init time, and then not being touched. Any code requiring locations is likely to get it once, then never look at it again. If `Override` is called later, that will never take effect. A nice option would be to have a static `Locations` type to get paths from, which can be overriden at creation time (but not changed later).  
The way it is in this PR I think it should work, because we only use `Override` in cmd/syncthing/main.go before the actual "syncthing service" is kicked off. Basically I don't think this PR should be blocked due to this, though it would of course be nice you would like to address the concern.

And wrapping on path separators would be awesome, but I am aware that's somewhere between super annoying and impossible, so will stay a dream :)

File: script/authors.go (hunk 5 above)
Comment (acolomb) [reply script/authors.go:120]: Yes it is, matching the container restructuring in that dialog.

File: lib/locations/locations.go (hunk 6 above)
Comment (acolomb) [reply lib/locations/locations.go:138]: Mostly the `${timestamp}` stuff which is expanded when creating such a file.  And I wanted to make sure whatever happens in other code, the returned paths are always real.  So yes, with the current usage pattern this should not actually filter out anything.

File: lib/locations/locations.go (hunk 7 above)
Comment (acolomb) [reply lib/locations/locations.go:142]: Also for the API? I think it makes the whole thing needlessly complicated when there is a `paths` property and then a separate `baseDirs` property, or one `paths` with two nested sub-objects (called what?).

All the keys in that map are well-defined strings, so this prefix is just a part of the keyspace.  It might look cleaner when making the whole thing a structure as you suggested.  Let me see what I can come up with there.  But do you think all the calling places should be changed to access a specific object instead of just calling a function from that module?  I'd rather not change that in principle.

**Comment by acolomb** [comment]
> Right now everything is set up at init time, and then not being touched. Any code requiring locations is likely to get it once, then never look at it again. If `Override` is called later, that will never take effect.

That's pretty much the same as the existing `SetBaseDir()` pattern, which overrides stuff from the command-line arguments or during tests.  Maybe we should just call it `SetLocation()` to make the similarity more obvious.  In general these two functions need to be called before any dependent components are initialized and read in their paths to use.

I'm not opposed to switching to a structure type of some kind, but I don't yet see how it will change the fundamental pattern of *parse args - override some locations - finalize (expand placeholders) - start services reading locations when needed*.  Open to more detailed suggestions.

File: lib/locations/locations.go (hunk 2 above)
Comment (acolomb) [reply lib/locations/locations.go]: I redid the whole function to be more aligned with `SetBaseDir()`, removing this restriction.

**Comment by acolomb** [comment]
Not sure what the Windows build failure is trying to tell me, I haven't touched the `autoClosedFile` stuff AFAICT.

File: lib/locations/locations.go (hunk 6 above)
Comment (imsodin) [reply lib/locations/locations.go:146]: Given it shouldn't happen, returning regardless of placeholders seems safer (and simpler). Then at least it's obvious that and what is wrong, while just a missing path is easy to miss.

File: lib/locations/locations.go (hunk 7 above)
Comment (imsodin) [reply lib/locations/locations.go:150]: lets leave as is

**Comment by imsodin** [comment]
> Not sure what the Windows build failure is trying to tell me, I haven't touched the `autoClosedFile` stuff AFAICT.

It's telling you nothing, it's flaky.

File: lib/locations/locations.go (hunk 6 above)
Comment (acolomb) [reply lib/locations/locations.go:146]: Okay, but I'll keep filtering the `${timestamp}` stuff, as those aren't really paths at any given time.

Comment (imsodin) [reply lib/locations/locations.go:146]: I don't care much honestly, but I don't see why? These paths are informational - and the panic-log path template is still useful information and a `${timestamp}` seems self-explantory enough to understand what it means to me.

**Comment by imsodin** [comment]
> I'm not opposed to switching to a structure type of some kind, but I don't yet see how it will change the fundamental pattern of parse args - override some locations - finalize (expand placeholders) - start services reading locations when needed. Open to more detailed suggestions.

If the content is private, you can set it once in the constructor/`New...` function - then no-one can touch it again. Thus the pattern you describe is enforced, no-one can later one come in and do a `Set...` later on.  
But with what you pointed out, even less reason to take this on in this PR. The problem already exists. So even if it would be nice if it was improved, shouldn't be mixed with the changes happening in this PR.

File: lib/locations/locations.go (hunk 6 above)
Comment (acolomb) [reply lib/locations/locations.go]: I guess it made more sense when the results of that API call were displayed as a whole in the GUI dialog.  Now that we pick explicit values in the GUI code with more elaborate descriptions, it's fine to have them in the API response and just not include them in the dialog (because non-tech-savvy users maybe won't find it self-explanatory).

**Comment by imsodin** [comment]
> I think this is done, please review. Do we want to keep the `paths` property in `/rest/system/status` as well or only on its own endpoint?

What's the purpose of the separate endpoint?  
Also what's the need to refresh paths when opening the modal view? The entire system infromation is refreshed every 10s, and this isn't information that changes often - if there's any controls in the web UI that can change them, I'd rather add a refresh call to them instead.

**Comment by acolomb** [comment]
> What's the purpose of the separate endpoint?

Exactly because the info never changes during runtime, we don't really need it as part of the 10 second refresh cycle.  However, the `tilde` property for example is also contained there, so that was the first logical choice.  Having a separate endpoint (*instead* of including it in `/rest/system/status`) allows to get the information only when needed, cutting down on the amount of cyclically refreshed data.

So you're suggesting to drop the separate endpoint and use only system status instead?  I'd actually prefer the opposite, making endpoints more granular instead of stuffing everything together into one big "status".

**Comment by imsodin** [comment]
I was questioning having it in both. Personally I'd probably go for simplicity over granularity/optimisation given the existing `/rest/system` endpoint (containing static info already), but I don't feel strongly either way. Just no duplication of `paths` please.

**Comment by acolomb** [comment]
I removed it from `/rest/system/status` again.  Having static information there seems unfortunate to me, so I avoid proliferating that.

**Comment by acolomb** [comment]
Ping. Anything else to get this merge-ready?

**Comment by acolomb** [comment]
Ping again.  Another week, another conflict resolution.  Kindly asking to get this reviewed / merged soonish.

**Comment by tomasz1986** [comment]
> Ping again. Another week, another conflict resolution. Kindly asking to get this reviewed / merged soonish.

I can only say that on the GUI side, everything seems to work perfectly fine. I'd just like to ask to maybe add a new screenshot, as the final version is quite different than what's being shown now.

Also, I'd like to ask a question regarding the "user home" terminology. Wouldn't it be better to make it more explicit, e.g. "user home directory" or "user home folder"? There's no "user home" on some operating systems, either because the terminology is different (i.e. Windows, which uses "user profile"), or because the concept itself doesn't exist (i.e. Android). I think at least adding the word "directory" or "folder" makes it easier to understand.

**Review by calmh — APPROVED** [review]

**Review by calmh — APPROVED** [review]

**Comment by acolomb** [comment]
> I think at least adding the word "directory" or "folder" makes it easier to understand.

Well it's under a "paths" heading, so that should point in the right direction. Anyway thanks for merging @calmh.
